chronos/timer.nim

```diff
@@ -8,45 +8,397 @@
+      ## Procedure's resolution is millisecond.
+      var t: Timeval
+      posix_gettimeofday(t)
+      result = (cast[uint64](t.tv_sec) * 1_000 +
+                cast[uint64](t.tv_usec) div 1_000)
 
-  proc posix_gettimeofday(tp: var Timeval, unused: pointer = nil) {.
-    importc: "gettimeofday", header: "<sys/time.h>".}
+    proc fastEpochTimeNano(): uint64 {.inline.} =
+      ## Procedure's resolution is nanosecond.
+      var t: Timeval
+      posix_gettimeofday(t)
+      result = (cast[uint64](t.tv_sec) * 1_000_000_000 +
+                cast[uint64](t.tv_usec) * 1_000)
+  else:
+    from posix import clock_gettime, Timespec, CLOCK_MONOTONIC
 
-  proc fastEpochTime*(): uint64 {.inline.} =
-    var t: Timeval
-    posix_gettimeofday(t)
-    result = (uint64(t.tv_sec) * 1_000 + uint64(t.tv_usec) div 1_000)
+    proc fastEpochTime*(): uint64 {.
+         inline, deprecated: "Use Moment.now()".} =
+      ## Procedure's resolution is millisecond.
+      var t: Timespec
+      discard clock_gettime(CLOCK_MONOTONIC, t)
+      result = ((cast[uint64](t.tv_sec) * 1_000) +
+                (cast[uint64](t.tv_nsec) div 1_000_000))
+
+    proc fastEpochTimeNano(): uint64 {.inline.} =
+      ## Procedure's resolution is nanosecond.
+      var t: Timespec
+      discard clock_gettime(CLOCK_MONOTONIC, t)
+      result = cast[uint64](t.tv_sec) * 1_000_000_000'u64 +
+               cast[uint64](t.tv_nsec)
 
 elif defined(posix):
+  from posix import clock_gettime, Timespec, CLOCK_REALTIME, CLOCK_MONOTONIC
+
+  when asyncTimer == "system":
+    proc fastEpochTime*(): uint64 {.
+         inline, deprecated: "Use Moment.now()".} =
+      ## Procedure's resolution is millisecond.
+      var t: Timespec
+      discard clock_gettime(CLOCK_REALTIME, t)
+      result = (cast[uint64](t.tv_sec) * 1_000 +
+                (cast[uint64](t.tv_nsec) div 1_000_000))
+
+    proc fastEpochTimeNano(): uint64 {.inline.} =
+      ## Procedure's resolution is nanosecond.
+      var t: Timespec
+      discard clock_gettime(CLOCK_REALTIME, t)
+      result = cast[uint64](t.tv_sec) * 1_000_000_000'u64 +
+               cast[uint64](t.tv_nsec)
 
-  from posix import clock_gettime, Timespec, CLOCK_REALTIME
+  else:
+    proc fastEpochTime*(): uint64 {.
+         inline, deprecated: "Use Moment.now()".} =
+      ## Procedure's resolution is millisecond.
+      var t: Timespec
+      discard clock_gettime(CLOCK_MONOTONIC, t)
+      result = (cast[uint64](t.tv_sec) * 1_000 +
+                (cast[uint64](t.tv_nsec) div 1_000_000))
 
-  proc fastEpochTime*(): uint64 {.inline.} =
-    var t: Timespec
-    discard clock_gettime(CLOCK_REALTIME, t)
-    result = (uint64(t.tv_sec) * 1_000 + uint64(t.tv_nsec) div 1_000_000)
+    proc fastEpochTimeNano(): uint64 {.inline.} =
+      ## Procedure's resolution is nanosecond.
+      var t: Timespec
+      discard clock_gettime(CLOCK_MONOTONIC, t)
+      result = cast[uint64](t.tv_sec) * 1_000_000_000'u64 +
```

Comment:
why cast and not just `uint64(t.tv_sec)`?

Reply:
Because `cast[T]` do not perform any checking, so its faster.